util/gen_effect_id.ts

```diff
@@ -34,6 +35,9 @@ type OutputEffectId = {
   [name: string]: string; // the id is converted to hex, so use string
 };
 
+const _SCRIPT_NAME = path.basename(import.meta.url);
+let log: ConsoleLogger;
```

Comment:
This feels like a bit of a code smell, to store the variable top-level like this but not assign it until we're in the default export function.

Is there a reason to not just `const log = new ConsoleLogger(); log.setLogLevel(<some sane default>)` here and then just override the log level later instead?

Reply:
> This feels like a bit of a code smell, to store the variable top-level like this but not assign it until we're in the default export function.
> 
> Is there a reason to not just `const log = new ConsoleLogger(); log.setLogLevel(<some sane default>)` here and then just override the log level later instead?

Yeah, that works.  I obviously didn't want to stick the variable definition in the export func because then I'd have to pass it down through all of the cascading funcs.  But there's no reason it can't be fully initialized with a default value.

Reply:
> This feels like a bit of a code smell, to store the variable top-level like this but not assign it until we're in the default export function.
> 
> Is there a reason to not just `const log = new ConsoleLogger(); log.setLogLevel(<some sane default>)` here and then just override the log level later instead?

Hmm.  That doesn't work.  Node throws an error on the new declaration when `npm run util` is used.
`ReferenceError: Cannot access 'ConsoleLogger' before initialization`

I'm wondering if maybe it's somehow a circular dependency problem, since making the call inside the export func works fine.  But I'm not sure how to address that.

Reply:
Yeah, it's circular dependencies. This diff should fix the issue. It also moves the default log level to a static readonly on the class instead of leaving it as a separate thing to import.

```diff
diff --git a/util/console_logger.ts b/util/console_logger.ts
new file mode 100644
index 000000000..07c0f550e
--- /dev/null
+++ b/util/console_logger.ts
@@ -0,0 +1,71 @@
+export const logLevels = [
+  ['alert', 'ALERTS: issues that probably require action (RECOMMENDED)'],
+  ['info', 'INFO: more routine/known issues that may not require action'],
+  ['debug', 'DEBUG: detailed output for troubleshooting'],
+  ['silent', 'SILENT: success & fatal errors only (not recommended)'],
+] as const;
+
+export type LogLevelKey = typeof logLevels[number][0];
+export type LogLevelLabel = typeof logLevels[number][1];
+
+const logLevelDefault: LogLevelKey = 'alert';
+
+export class ConsoleLogger {
+  public static readonly logLevelDefault: LogLevelKey = 'alert';
+
+  // assign numerical values to log levels so we can do a quick compare
+  // in deciding whether a user wants to see that type of log output
+  logLevelMap: { [K in LogLevelKey]: 0 | 1 | 2 | 3 } = {
+    silent: 0,
+    alert: 1,
+    info: 2,
+    debug: 3,
+  };
+  myLogLevel: typeof this.logLevelMap[LogLevelKey];
+
+  setLogLevel(logLevel?: LogLevelKey): void {
+    // class is initialized in scripts outside of all constructs,
+    // so it doesn't need to be passed to every function
+    // this happens before the default function is called
+    // with the user's log level choice.
+    if (logLevel === undefined)
+      return;
+    if (!Object.keys(this.logLevelMap).includes(logLevel))
+      this.fatalError('Invalid log level is set.');
+    this.myLogLevel = this.logLevelMap[logLevel];
+    this.debug(`Log level set: ${logLevel}`);
+  }
+
+  constructor(userLogLevelKey?: LogLevelKey) {
+    this.myLogLevel = this.logLevelMap[logLevelDefault]; // needs to be initialized first
+    this.setLogLevel(userLogLevelKey);
+  }
+
+  alert(msg: string): void {
+    if (this.myLogLevel >= this.logLevelMap.alert)
+      console.log(`Alert: ${msg}`);
+  }
+
+  info(msg: string): void {
+    if (this.myLogLevel >= this.logLevelMap.info)
+      console.log(`Info: ${msg}`);
+  }
+
+  debug(msg: string): void {
+    if (this.myLogLevel >= this.logLevelMap.debug)
+      console.log(`Debug: ${msg}`);
+  }
+
+  successDone(msg: string): void {
+    console.log(`Success: ${msg}`);
+  }
+
+  nonFatalError(msg: string): void {
+    console.log(`ERROR: ${msg}`);
+  }
+
+  fatalError(msg: string): void {
+    console.log(`ERROR: ${msg} Exiting...`);
+    process.exit(1);
+  }
+}
diff --git a/util/gen_effect_id.ts b/util/gen_effect_id.ts
index 75c240f1c..5eba3d958 100644
--- a/util/gen_effect_id.ts
+++ b/util/gen_effect_id.ts
@@ -1,7 +1,7 @@
 import path from 'path';
 
+import { ConsoleLogger, LogLevelKey } from './console_logger';
 import { cleanName } from './csv_util';
-import { ConsoleLogger, LogLevelKey } from './generate_data_files';
 import { OutputFileAttributes, XivApi } from './xivapi';
 
 const _EFFECT_ID: OutputFileAttributes = {
@@ -36,7 +36,8 @@ type OutputEffectId = {
 };
 
 const _SCRIPT_NAME = path.basename(import.meta.url);
-let log: ConsoleLogger;
+const log = new ConsoleLogger();
+log.setLogLevel('alert');
 
 // TODO: add renaming?
 // Almagest: 563
@@ -162,7 +163,6 @@ const assembleData = (apiData: XivApiStatus): OutputEffectId => {
 };
 
 export default async (logLevel: LogLevelKey): Promise<void> => {
-  log = new ConsoleLogger();
   log.setLogLevel(logLevel);
   log.info(`Starting processing for ${_SCRIPT_NAME}`);
 
diff --git a/util/gen_hunt_data.ts b/util/gen_hunt_data.ts
index 39956b1a0..5a145e9c1 100644
--- a/util/gen_hunt_data.ts
+++ b/util/gen_hunt_data.ts
@@ -2,8 +2,8 @@ import path from 'path';
 
 import { LocaleObject } from '../types/trigger';
 
+import { ConsoleLogger, LogLevelKey } from './console_logger';
 import { getCnTable, getKoTable } from './csv_util';
-import { ConsoleLogger, LogLevelKey } from './generate_data_files';
 import { OutputFileAttributes, XivApi } from './xivapi';
 
 const _HUNT: OutputFileAttributes = {
diff --git a/util/gen_pet_names.ts b/util/gen_pet_names.ts
index b31ac5269..795348fa3 100644
--- a/util/gen_pet_names.ts
+++ b/util/gen_pet_names.ts
@@ -1,7 +1,7 @@
 import path from 'path';
 
+import { ConsoleLogger, LogLevelKey } from './console_logger';
 import { getCnTable, getKoTable } from './csv_util';
-import { ConsoleLogger, LogLevelKey } from './generate_data_files';
 import { OutputFileAttributes, XivApi } from './xivapi';
 
 const _PET_NAMES: OutputFileAttributes = {
diff --git a/util/gen_weather_rate.ts b/util/gen_weather_rate.ts
index bb3e0af45..ed24c75d1 100644
--- a/util/gen_weather_rate.ts
+++ b/util/gen_weather_rate.ts
@@ -1,6 +1,6 @@
 import path from 'path';
 
-import { ConsoleLogger, LogLevelKey } from './generate_data_files';
+import { ConsoleLogger, LogLevelKey } from './console_logger';
 import { OutputFileAttributes, XivApi } from './xivapi';
 
 const _WEATHER_RATE: OutputFileAttributes = {
diff --git a/util/gen_world_ids.ts b/util/gen_world_ids.ts
index 4ebcf0cd0..c37c815ca 100644
--- a/util/gen_world_ids.ts
+++ b/util/gen_world_ids.ts
@@ -1,13 +1,13 @@
 import path from 'path';
 
-import { ConsoleLogger, LogLevelKey } from './generate_data_files';
+import { ConsoleLogger, LogLevelKey } from './console_logger';
 import { OutputFileAttributes, XivApi } from './xivapi';
 
 const _WORLD_ID: OutputFileAttributes = {
   outputFile: 'resources/world_id.ts',
   type: 'Worlds',
   header: `// NOTE: This data is filtered to public worlds only (i.e. isPublic: true)
-  
+
   export type DataCenter = {
     id: number;
     name: string;
diff --git a/util/gen_zone_id_and_info.ts b/util/gen_zone_id_and_info.ts
index 823f48f78..4f4c0a258 100644
--- a/util/gen_zone_id_and_info.ts
+++ b/util/gen_zone_id_and_info.ts
@@ -2,8 +2,8 @@ import path from 'path';
 
 import { LocaleText } from '../types/trigger';
 
+import { ConsoleLogger, LogLevelKey } from './console_logger';
 import { cleanName, getCnTable, getKoTable } from './csv_util';
-import { ConsoleLogger, LogLevelKey } from './generate_data_files';
 import { OutputFileAttributes, XivApi } from './xivapi';
 import Overrides from './zone_overrides';
 
diff --git a/util/generate_data_files.ts b/util/generate_data_files.ts
index 5a8dc1e30..1a9b19c0b 100644
--- a/util/generate_data_files.ts
+++ b/util/generate_data_files.ts
@@ -3,6 +3,7 @@ import inquirer from 'inquirer';
 
 import { UnreachableCode } from '../resources/not_reached';
 
+import { ConsoleLogger, LogLevelKey, LogLevelLabel, logLevels } from './console_logger';
 import { default as generateEffectIds } from './gen_effect_id';
 import { default as generateHunt } from './gen_hunt_data';
 import { default as generatePetNames } from './gen_pet_names';
@@ -24,13 +25,6 @@ const fileKeys = [
 const allKey = 'all';
 const allLabel = '* Generate All Data Files';
 
-const logLevels = [
-  ['alert', 'ALERTS: issues that probably require action (RECOMMENDED)'],
-  ['info', 'INFO: more routine/known issues that may not require action'],
-  ['debug', 'DEBUG: detailed output for troubleshooting'],
-  ['silent', 'SILENT: success & fatal errors only (not recommended)'],
-] as const;
-
 type FileKey = typeof fileKeys[number];
 
 const fileKeyToFunc: { [K in FileKey]: (logLevel: LogLevelKey) => Promise<void> } = {
@@ -61,9 +55,6 @@ const fileChoices: { name: FileLabelAll; value: FileKeyAll }[] = [
 const fileKeyToFuncAll: { [filename in FileKeyAll]: (logLevel: LogLevelKey) => Promise<void> } =
   ({ [allKey]: generateAll, ...fileKeyToFunc });
 
-export type LogLevelKey = typeof logLevels[number][0];
-type LogLevelLabel = typeof logLevels[number][1];
-
 const logLevelChoices: { name: LogLevelLabel; value: LogLevelKey }[] = logLevels.map((ll) => ({
   name: ll[1],
   value: ll[0],
@@ -83,7 +74,6 @@ type GenerateDataFilesInquirerType = {
   [name in keyof GenerateDataFilesNamespaceInterface]: GenerateDataFilesNamespaceInterface[name];
 };
 
-const logLevelDefault: LogLevelKey = 'alert';
 const fileDefault: FileKeyAll = 'all';
 
 // TODO: argparse isn't handling CLI options past the initial 'action'
@@ -112,7 +102,7 @@ const generateDataFilesFunc = async (args: Namespace): Promise<void> => {
   return inquirer.prompt<GenerateDataFilesInquirerType>(questions)
     .then((answers) => {
       const myChoice: FileKeyAll = answers.file ?? args.file ?? fileDefault;
-      const myLogLevel: LogLevelKey = answers.loglevel ?? args.loglevel ?? logLevelDefault;
+      const myLogLevel = answers.loglevel ?? args.loglevel ?? ConsoleLogger.logLevelDefault;
       return fileKeyToFuncAll[myChoice](myLogLevel);
     }).catch(console.error);
 };
@@ -144,61 +134,3 @@ export const registerGenerateDataFiles = (
     help: 'The level of console output you want to see',
   });
 };
-
-export class ConsoleLogger {
-  // assign numerical values to log levels so we can do a quick compare
-  // in deciding whether a user wants to see that type of log output
-  logLevelMap: { [K in LogLevelKey]: 0 | 1 | 2 | 3 } = {
-    silent: 0,
-    alert: 1,
-    info: 2,
-    debug: 3,
-  };
-  myLogLevel: typeof this.logLevelMap[LogLevelKey];
-
-  setLogLevel(logLevel?: LogLevelKey): void {
-    // class is initialized in scripts outside of all constructs,
-    // so it doesn't need to be passed to every function
-    // this happens before the default function is called
-    // with the user's log level choice.
-    if (logLevel === undefined)
-      return;
-    if (!Object.keys(this.logLevelMap).includes(logLevel))
-      this.fatalError('Invalid log level is set.');
-    this.myLogLevel = this.logLevelMap[logLevel];
-    this.debug(`Log level set: ${logLevel}`);
-  }
-
-  constructor(userLogLevelKey?: LogLevelKey) {
-    this.myLogLevel = this.logLevelMap[logLevelDefault]; // needs to be initialized first
-    this.setLogLevel(userLogLevelKey);
-  }
-
-  alert(msg: string): void {
-    if (this.myLogLevel >= this.logLevelMap.alert)
-      console.log(`Alert: ${msg}`);
-  }
-
-  info(msg: string): void {
-    if (this.myLogLevel >= this.logLevelMap.info)
-      console.log(`Info: ${msg}`);
-  }
-
-  debug(msg: string): void {
-    if (this.myLogLevel >= this.logLevelMap.debug)
-      console.log(`Debug: ${msg}`);
-  }
-
-  successDone(msg: string): void {
-    console.log(`Success: ${msg}`);
-  }
-
-  nonFatalError(msg: string): void {
-    console.log(`ERROR: ${msg}`);
-  }
-
-  fatalError(msg: string): void {
-    console.log(`ERROR: ${msg} Exiting...`);
-    process.exit(1);
-  }
-}
diff --git a/util/xivapi.ts b/util/xivapi.ts
index fb5d4cae3..1726f7e24 100644
--- a/util/xivapi.ts
+++ b/util/xivapi.ts
@@ -9,7 +9,7 @@ import { exec } from '@actions/exec';
 import eslint from 'eslint';
 import fetch from 'node-fetch';
 
-import { ConsoleLogger } from './generate_data_files';
+import { ConsoleLogger } from './console_logger';
 
 const _XIVAPI_URL = 'https://xivapi.com/';
 

```

Reply:
That's great.  Thanks for the assist!